libdevcrypto/OverlayDB.h

```diff
@@ -33,28 +34,39 @@ namespace dev
 class OverlayDB: public MemoryDB
 {
 public:
-	OverlayDB(ldb::DB* _db = nullptr): m_db(_db) {}
+	OverlayDB(ldb::DB* _db = nullptr, unsigned _pruning = 0): m_db(_db), m_pruning(_pruning) {}
 	~OverlayDB();
 
 	ldb::DB* db() const { return m_db.get(); }
 
-	void commit();
+	void commit(u256 _blockNumber);
 	void rollback();
+	void insert(h256 const& _h, bytesConstRef _v);
 
 	std::string lookup(h256 const& _h) const;
 	bool exists(h256 const& _h) const;
 	void kill(h256 const& _h);
-	bool deepkill(h256 const& _h);
 
 	bytes lookupAux(h256 const& _h) const;
 
 private:
+	void safeWrite(ldb::WriteBatch& _batch) const;
 	using MemoryDB::clear;
 
 	std::shared_ptr<ldb::DB> m_db;
 
 	ldb::ReadOptions m_readOptions;
 	ldb::WriteOptions m_writeOptions;
+
+	unsigned m_pruning;
+
+	u256 isInDeathRow(h256 const& _h) const;
+	int getRefCount(h256 const& _h) const;
+	int increaseRefCount(h256 const& _h, ldb::WriteBatch& _batch, int _addedRefCount = 1, bool _revert = false) const;
+
+	static std::map<u256, std::set<h256> > m_deathrow;
+	static std::map<u256, std::unordered_map<h256, int > > m_changes;
```

Comment:
This should be disk backed. Otherwise you'll end up with wrong counters if chain reorg happens right after start.